semantic/src/main/kotlin/tools/samt/semantic/checks/ValidTypeExpressionCheck.kt

```diff
@@ -0,0 +1,155 @@
+                    checkConstraintArgumentExpression(argument)
+                }
+            }
+
+            is GenericSpecializationNode -> diagnostics.reportError(
+                "Generic type are currently not supported",
+                expression.location
+            )
+
+            is BooleanNode,
+            is NumberNode,
+            is StringNode,
+            -> diagnostics.reportError("Cannot use literal value as type", expression.location)
+
+            is ObjectNode,
+            is ArrayNode,
+            is RangeExpressionNode,
+            is WildcardNode,
+            -> diagnostics.reportError("Invalid type expression", expression.location)
+        }
+    }
+
+    private fun checkConstraintBaseTypeExpression(expression: ExpressionNode, optionalAllowed: Boolean = true) {
+        when (expression) {
+            is IdentifierNode,
+            is BundleIdentifierNode,
+            -> Unit
+
+            is OptionalDeclarationNode -> {
+                if (optionalAllowed) {
+                    checkConstraintBaseTypeExpression(expression.base, false)
+                } else {
+                    diagnostics.reportError("Cannot nest optional types", expression.location)
+                }
+            }
+
+            is CallExpressionNode -> {
+                diagnostics.reportError("Cannot nest constraint types", expression.location)
+            }
+
+            is GenericSpecializationNode -> diagnostics.reportError(
+                "Generic type are currently not supported",
+                expression.location
+            )
+
+            is BooleanNode,
+            is NumberNode,
+            is StringNode,
+            is ObjectNode,
+            is ArrayNode,
+            is RangeExpressionNode,
+            is WildcardNode,
+            -> diagnostics.reportError("Base type for constraint is not valid", expression.location)
+        }
+    }
+
+    private fun checkConstraintArgumentExpression(expression: ExpressionNode) {
+        when (expression) {
+            is CallExpressionNode -> {
+                if (expression.base !is BundleIdentifierNode) {
+                    diagnostics.reportError("Constraint argument name must be an identifier", expression.base.location)
+                }
+                for (argument in expression.arguments) {
+                    checkLiteralExpression(argument)
+                }
+            }
+            is RangeExpressionNode -> checkRangeExpression(expression)
+
+            is BooleanNode,
+            is NumberNode,
+            is StringNode,
+            is IdentifierNode,
+            is BundleIdentifierNode,
+            is OptionalDeclarationNode,
+            is ObjectNode,
+            is ArrayNode,
+            is WildcardNode,
+            is GenericSpecializationNode,
+            -> diagnostics.reportError("Invalid constraint, examples of valid constraints are 'size(1..*)', 'pattern(\"a-z\")' and '1..*'", expression.location)
```

Comment:
Split this into a `reportError` which only reports the error and an additional `reportInfo` call to show hints as to what sort of constraints are legal.